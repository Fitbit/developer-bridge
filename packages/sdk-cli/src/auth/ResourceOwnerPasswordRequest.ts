export interface ResoureOwnerTokenRequestJson {
  grant_type: 'password';
  client_id: string;
  username: string;
  password: string;
  extras?: Record<string, string>;
}

export default class ResoureOwnerTokenRequest {
  clientId: string;
  username: string;
  password: string;
  extras?: Record<string, string>;

  constructor(request: ResoureOwnerTokenRequestJson) {
    this.clientId = request.client_id;
    this.username = request.username;
    this.password = request.password;
    this.extras = request.extras;
  }

  /**
   * Serializes a TokenRequest to a JavaScript object.
   */
  toJson(): ResoureOwnerTokenRequestJson {
    return {
      grant_type: 'password',
      username: this.username,
      password: this.password,
      client_id: this.clientId,
      extras: this.extras,
    };
  }

  toStringMap(): Record<string, string> {
    const map: Record<string, string> = {
      grant_type: 'password',
      client_id: this.clientId,
      username: this.username,
      password: this.password,
    };

    // copy over extras
    if (this.extras) {
      for (const extra in this.extras) {
        if (this.extras.hasOwnProperty(extra) && !map.hasOwnProperty(extra)) {
          // check before inserting to requestMap
          map[extra] = this.extras[extra];
        }
      }
    }
    return map;
  }
}
