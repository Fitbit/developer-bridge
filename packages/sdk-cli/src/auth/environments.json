{
  "production": {
    "clientId": "228TQF",
    "apiUrl": "https://api.fitbit.com:443",
    "devRelayUrl": "wss://027-v3-api-soa.fitbit.com:443/dbridge/"
  },
  "stage": {
    "clientId": "228VJK",
    "apiUrl": "https://api.fitbit-stage.com:443",
    "devRelayUrl": "wss://developer-relay-stage-frontend.site-ops.fitbit.com/"
  },
  "int": {
    "clientId": "228W5N",
    "apiUrl": "https://api.fitbit-int.com:443",
    "devRelayUrl": "wss://developer-relay-int-frontend.site-ops.fitbit.com/"
  }
}
