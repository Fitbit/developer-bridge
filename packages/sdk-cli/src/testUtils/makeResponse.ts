import { Response } from '../fetch';

export default function makeResponse(
  init: ResponseInit = { status: 200 },
  body = '{}',
) {
  const response = new Response(
    body,
    {
      statusText: `Status ${init.status}`,
      ...init,
    },
  );

  Object.defineProperty(response, 'url', {
    value: 'http://api',
    writable: false,
  });

  return response;
}
