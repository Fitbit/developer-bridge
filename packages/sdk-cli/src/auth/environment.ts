import process from 'process';
// tslint:disable-next-line:import-name
import environments from './environments.json';

export default () => {
  const defaultEnv = 'production';
  const environment = (
    process.env.FITBIT_SDK_ENVIRONMENT || defaultEnv
  ) as keyof typeof environments;

  if (Object.keys(environments).indexOf(environment) === -1) {
    throw new Error(`Invalid environment specified: ${environment}`);
  }

  const config = environments[environment];

  const clientId = process.env.FITBIT_SDK_CLIENT_ID;

  if (clientId) {
    config.clientId = clientId;
  }

  return {
    environment,
    config,
  };
};
