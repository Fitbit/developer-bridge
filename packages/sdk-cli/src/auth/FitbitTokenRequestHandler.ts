import {
  AuthorizationServiceConfiguration,
} from '@openid/appauth/built/authorization_service_configuration';
import { AppAuthError } from '@openid/appauth/built/errors';
import { RevokeTokenRequest } from '@openid/appauth/built/revoke_token_request';
import {
  ErrorType,
  TokenError,
  TokenResponse,
} from '@openid/appauth/built/token_response';
import {
  TokenRequestHandler,
} from '@openid/appauth/built/token_request_handler';
import { BasicQueryStringUtils } from '@openid/appauth/built/query_string_utils';
import * as t from 'io-ts';

import environment from './environment';
import fetch from '../fetch';

const { clientId, clientSecret } = environment().config;

// Fitbit's error responses are non-standard per
// https://tools.ietf.org/html/rfc6749#section-5.2
// tslint:disable-next-line:variable-name
const FitbitAuthErrorResponse = t.interface({
  errors: t.array(
    t.interface({
      errorType: t.string,
      message: t.string,
    }),
  ),
  success: t.boolean,
});
type FitbitAuthErrorResponse = t.TypeOf<typeof FitbitAuthErrorResponse>;

// tslint:disable-next-line:variable-name
const FitbitTokenResponse = t.interface({
  access_token: t.string,
  refresh_token: t.string,
  expires_in: t.number,
});
type FitbitTokenResponse = t.TypeOf<typeof FitbitTokenResponse>;

const commonParams = {
  method: 'POST',
  headers: {
    'Content-Type': 'application/x-www-form-urlencoded',
    Authorization: 'Basic ' + Buffer.from(`${clientId}:${clientSecret}`).toString('base64'),
  },
};

export default class FitbitTokenRequestHandler implements TokenRequestHandler {
  public readonly utils = new BasicQueryStringUtils();

  performRevokeTokenRequest(
    configuration: AuthorizationServiceConfiguration,
    request: RevokeTokenRequest,
  ): Promise<boolean> {
    return fetch(
      configuration.revocationEndpoint,
      {
        ...commonParams,
        body: this.utils.stringify({
          token: request.token,
        }),
      },
    ).then(response => response.ok);
  }

  async performTokenRequest(
    configuration: AuthorizationServiceConfiguration,
    request: {
      toStringMap: () => Record<string, string>,
    },
  ): Promise<TokenResponse> {
    const response = await fetch(
      configuration.tokenEndpoint,
      {
        ...commonParams,
        body: this.utils.stringify(request.toStringMap()),
      },
    );
    const responseJson = await response.json();

    if (response.status === 200 && FitbitTokenResponse.is(responseJson)) {
      return new TokenResponse(responseJson);
    }

    if (FitbitAuthErrorResponse.is(responseJson)) {
      const error = responseJson.errors[0];
      throw new AppAuthError(
        error.errorType,
        new TokenError({
          error: error.errorType as ErrorType,
          error_description: error.message,
        }),
      );
    }

    throw new AppAuthError(
      `Unexpected response format for status ${response.status}: ${JSON.stringify(responseJson)}`,
    );
  }
}
