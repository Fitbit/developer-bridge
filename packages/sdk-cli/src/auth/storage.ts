import { TokenResponse } from '@openid/appauth/built/token_response';
import * as t from 'io-ts';
import keytar from 'keytar';

import environment from './environment';

const keyStoreService = 'fitbit-sdk';
const keyStoreAccount = environment().environment;

// tslint:disable-next-line:variable-name
const AuthStorage = t.interface({
  access_token: t.string,
  refresh_token: t.string,
  issued_at: t.number,
  expires_in: t.string,
});
type AuthStorage = t.TypeOf<typeof AuthStorage>;

function set(tokenResponse: TokenResponse) {
  return keytar.setPassword(
    keyStoreService,
    keyStoreAccount,
    JSON.stringify(tokenResponse.toJson()),
  );
}

async function get(): Promise<TokenResponse | null> {
  const authDataStr = await keytar.getPassword(keyStoreService, keyStoreAccount);
  if (authDataStr === null) return null;
  let authData;
  try {
    authData = JSON.parse(authDataStr);
  } catch (ex) {
    await clear();
    return null;
  }
  if (!AuthStorage.is(authData)) {
    await clear();
    return null;
  }
  return new TokenResponse(authData);
}

function clear() {
  return keytar.deletePassword(keyStoreService, keyStoreAccount);
}

export default {
  set,
  get,
  clear,
};
