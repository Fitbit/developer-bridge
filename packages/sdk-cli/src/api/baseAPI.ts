import * as t from 'io-ts';
import { decode } from '@fitbit/jsonrpc-ts';

import * as auth from '../auth';
import environment from '../auth/environment';
import fetch, { Headers } from '../fetch';
import { assertJSON, okOrElse } from '../util/fetchUtil';

// tslint:disable-next-line:variable-name
const APIErrorResponse = t.type(
  {
    errors: t.array(t.type({
      message: t.string,
    })),
  },
  'Hosts403Response',
);
type APIErrorResponse = t.TypeOf<typeof APIErrorResponse>;

export const assertAPIResponseOK = okOrElse((response) => {
  return assertJSON()(response)
    .catch(() => Promise.reject(
      // tslint:disable-next-line:max-line-length
      new Error(`Fetch of ${response.url} returned status ${response.status} ${response.statusText}`),
    ))
    .then(responseObject => Promise.reject(new Error(APIErrorResponse.decode(responseObject).fold(
      // tslint:disable-next-line:max-line-length
      () => `fetch of ${response.url} returned status ${response.status} ${response.statusText} with body: ${JSON.stringify(responseObject, undefined, 2)}`,
      errorObj => errorObj.errors.map(err => err.message).join('\n'),
    )))
    .then(() => Promise.resolve(response)));
});

export const decodeJSON = <A, O, I>(endpointType: t.Type<A, O, I>) =>
  (response: Response): Promise<A> =>
    assertAPIResponseOK(response)
      .then(assertJSON())
      .then(decode(endpointType));

export async function apiFetch(path: string, init: RequestInit = {}) {
  const authToken = await auth.getAccessToken();
  if (!authToken) throw new Error(`Fetch of ${path} failed: no stored auth token`);

  const headers = new Headers(init.headers || {});
  headers.set('authorization', `Bearer ${authToken}`);
  const mergedInit: RequestInit = {
    cache: 'no-store',
    ...init,
    headers,
  };
  return fetch(`${environment().config.apiUrl}/${path}`, mergedInit);
}
