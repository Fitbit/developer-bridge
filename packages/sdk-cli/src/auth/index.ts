import { AuthorizationRequest } from '@openid/appauth/built/authorization_request';
import {
  AuthorizationNotifier,
  AuthorizationRequestHandler,
} from '@openid/appauth/built/authorization_request_handler';
import { AuthorizationResponse } from '@openid/appauth/built/authorization_response';
import {
  AuthorizationServiceConfiguration,
} from '@openid/appauth/built/authorization_service_configuration';
import * as appAuthFlags from '@openid/appauth/built/flags';
import { NodeBasedHandler, NodeCrypto } from '@openid/appauth/built/node_support';
import { RevokeTokenRequest } from '@openid/appauth/built/revoke_token_request';
import {
  GRANT_TYPE_AUTHORIZATION_CODE,
  GRANT_TYPE_REFRESH_TOKEN,
  TokenRequest,
} from '@openid/appauth/built/token_request';

import environment from './environment';
import storage from './storage';
import FitbitTokenRequestHandler from './FitbitTokenRequestHandler';

// TODO: PR against AppAuth to make this properly configurable
(appAuthFlags as any).IS_LOG = false;

const tokenHandler = new FitbitTokenRequestHandler();

function getAuthConfiguration() {
  const { apiUrl } = environment().config;
  return new AuthorizationServiceConfiguration({
    token_endpoint: `${apiUrl}/oauth2/token`,
    authorization_endpoint: `${apiUrl}/oauth2/authorize`,
    revocation_endpoint: `${apiUrl}/oauth2/revoke`,
  });
}

export async function refreshToken(refreshToken: string) {
  const { clientId } = environment().config;
  try {
    const response = await tokenHandler.performTokenRequest(
      getAuthConfiguration(),
      new TokenRequest({
        client_id: clientId,
        redirect_uri: '', // not needed
        grant_type: GRANT_TYPE_REFRESH_TOKEN,
        refresh_token: refreshToken,
      }),
    );
    await storage.set(response);
    return response.accessToken;
  } catch (ex) {
    await storage.clear();
    throw ex;
  }
}

function authError(msg: string) {
  return new Error(`Authorization error: ${msg}`);
}

function authorizationCallbackPromise(handler: AuthorizationRequestHandler) {
  const notifier = new AuthorizationNotifier();
  handler.setAuthorizationNotifier(notifier);

  return new Promise<{
    request: AuthorizationRequest,
    response: AuthorizationResponse,
  }>((resolve, reject) => {
    notifier.setAuthorizationListener((request, response, error) => {
      if (error) {
        if (error.errorDescription) {
          return reject(
            authError(`${error.error}: ${error.errorDescription}`),
          );
        }
        return reject(authError(error.error));
      }
      resolve({ request: request!, response: response! });
    });
  });
}

async function authorize() {
  const { clientId } = environment().config;

  const port = 13579;
  const redirectUri = `http://127.0.0.1:${port}`;

  const authorizationHandler = new NodeBasedHandler(port);
  authorizationHandler.performAuthorizationRequest(
    getAuthConfiguration(),
    new AuthorizationRequest(
      {
        client_id: clientId,
        redirect_uri: redirectUri,
        scope: 'profile',
        response_type: AuthorizationRequest.RESPONSE_TYPE_CODE,
      },
      new NodeCrypto(),
      true,
    ),
  );

  const { request, response } = await authorizationCallbackPromise(authorizationHandler);
  if (request.state !== response.state) throw authError('Mismatched state');
  return {
    redirectUri,
    code: response.code,
    pkceVerifier: request.internal!['code_verifier'],
  };
}

async function revoke(token: string) {
  await tokenHandler.performRevokeTokenRequest(
    getAuthConfiguration(),
    new RevokeTokenRequest({ token }),
  );
}

export async function login() {
  const { clientId } = environment().config;

  const { code, pkceVerifier, redirectUri } = await authorize();
  const response = await tokenHandler.performTokenRequest(
    getAuthConfiguration(),
    new TokenRequest({
      code,
      client_id: clientId,
      redirect_uri: redirectUri,
      grant_type: GRANT_TYPE_AUTHORIZATION_CODE,
      extras: { code_verifier: pkceVerifier },
    }),
  );
  await storage.set(response);
}

export async function logout() {
  const authData = await storage.get();
  if (!authData) return;
  await Promise.all([
    revoke(authData.accessToken),
    storage.clear(),
  ]);
}

export async function getAccessToken() {
  const authData = await storage.get();
  if (authData === null) return null;
  // Check for validity without any time buffer.
  if (authData.isValid(0)) return authData.accessToken;
  return refreshToken(authData.refreshToken!);
}
