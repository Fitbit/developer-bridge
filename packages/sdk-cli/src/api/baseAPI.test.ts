import nock from 'nock';

import * as auth from '../auth';
import * as fetchModule from '../fetch';

import environment from '../auth/environment';
import * as baseAPI from './baseAPI';
import makeResponse from '../testUtils/makeResponse';

jest.mock('../auth');

const mockAccessTokenContent = 'mockAccess';
const fakeAPIPath = 'fakeAPI';

function mockAuthToken(token: any = mockAccessTokenContent) {
  (auth.getAccessToken as jest.Mock).mockResolvedValueOnce(token);
}

describe('apiFetch()', () => {
  it('rejects if no auth header is available', () => {
    mockAuthToken(null);
    return expect(
      baseAPI.apiFetch(fakeAPIPath),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('can use a custom API URL and skip auth', async () => {
    jest
      .spyOn(fetchModule, 'default')
      .mockImplementationOnce(
        async (url: RequestInfo, init?: RequestInit) =>
          ({ url, headers: init?.headers } as Response),
      );

    const fakeAPIDomain = 'https://fake-dev-relay-test-endpoint.fitbit.com';

    mockAuthToken(null);
    // apiFetch would throw if it didn't skip auth, because authToken is falsy
    const response = await baseAPI.apiFetch(
      fakeAPIPath,
      undefined,
      fakeAPIDomain,
      false,
    );

    expect(response).toBeDefined();
    expect(auth.getAccessToken).toBeCalledTimes(0);
    expect(response.url).toBe(`${fakeAPIDomain}/${fakeAPIPath}`);
    expect(response.headers.has('authorization')).toBe(false);
  });

  describe('when an auth token is available', () => {
    let endpointScope: nock.Scope;
    beforeEach(() => {
      endpointScope = nock(environment().config.apiUrl, {
        reqheaders: {
          Authorization: `Bearer ${mockAccessTokenContent}`,
        },
      });

      it.each(['GET', 'POST'])(
        'sends an authorization header with requests where method is %s',
        async (method) => {
          const endpointCall = endpointScope
            .intercept(fakeAPIPath, method)
            .reply(200);

          mockAuthToken();
          await baseAPI.apiFetch(fakeAPIPath, { method });
          expect(endpointCall.isDone()).toBe(true);
        },
      );
    });
  });
});

describe('assertAPIResponseOK()', () => {
  it.each([400, 401, 403, 500, 502, 503])(
    'rejects on status code %d with a non-JSON body',
    (status) => {
      return expect(
        baseAPI.assertAPIResponseOK(makeResponse({ status })),
      ).rejects.toThrowErrorMatchingSnapshot();
    },
  );

  it('resolves to a response object on status code 200', () => {
    return expect(
      baseAPI.assertAPIResponseOK(makeResponse()),
    ).resolves.toBeInstanceOf(fetchModule.Response);
  });

  it('parses an error response with a valid JSON body', () => {
    return expect(
      baseAPI.assertAPIResponseOK(
        makeResponse(
          { headers: { 'Content-Type': 'application/json' }, status: 403 },
          JSON.stringify({ errors: [{ message: 'some error' }] }),
        ),
      ),
    ).rejects.toThrowError('some error');
  });

  it('throws an error containing the response body when the error JSON is invalid', () => {
    return expect(
      baseAPI.assertAPIResponseOK(
        makeResponse(
          { headers: { 'Content-Type': 'application/json' }, status: 403 },
          JSON.stringify({ message: 'some error' }),
        ),
      ),
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
