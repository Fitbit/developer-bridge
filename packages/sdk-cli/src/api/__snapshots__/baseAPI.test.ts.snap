// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiFetch() rejects if no auth header is available 1`] = `"Fetch of fakeAPI failed: no stored auth token"`;

exports[`assertAPIResponseOK() rejects on status code 400 with a non-JSON body 1`] = `"Fetch of http://api returned status 400 Status 400"`;

exports[`assertAPIResponseOK() rejects on status code 401 with a non-JSON body 1`] = `"Fetch of http://api returned status 401 Status 401"`;

exports[`assertAPIResponseOK() rejects on status code 403 with a non-JSON body 1`] = `"Fetch of http://api returned status 403 Status 403"`;

exports[`assertAPIResponseOK() rejects on status code 500 with a non-JSON body 1`] = `"Fetch of http://api returned status 500 Status 500"`;

exports[`assertAPIResponseOK() rejects on status code 502 with a non-JSON body 1`] = `"Fetch of http://api returned status 502 Status 502"`;

exports[`assertAPIResponseOK() rejects on status code 503 with a non-JSON body 1`] = `"Fetch of http://api returned status 503 Status 503"`;

exports[`assertAPIResponseOK() throws an error containing the response body when the error JSON is invalid 1`] = `
"fetch of http://api returned status 403 Status 403 with body: {
  \\"message\\": \\"some error\\"
}"
`;
